### PR TITLE
feat: add @auth/astro framework library with example & docs

### PR DESCRIPTION
**Pull request by theotterlord**
> _NOTE_:
>
> - It's a good idea to open an issue first to discuss potential changes.
> - Please make sure that you are _NOT_ opening a PR to fix a potential security vulnerability. Instead, please follow the [Security guidelines](https://github.com/nextauthjs/.github/blob/main/SECURITY.md) to disclose the issue to us confidentially.

## ☕️ Reasoning

**Supersedes #6425**

:rocket: Introduces an Astro framework package, with documentation & an example. The package exposes Astro components in addition to the methods.

Thanks to @tajkirkpatrick and @nowaythatworked who've helped out with building, debugging & documenting this package.

I've deployed the included example application to https://auth-astro-example.vercel.app/ feel free to take a look.

## 🧢 Checklist

- [x] Documentation
- [ ] Tests
- [x] Ready to be merged

## 🎫 Affected issues

Please [scout and link issues](https://github.com/nextauthjs/next-auth/issues) that might be solved by this PR.

Fixes: INSERT_ISSUE_LINK_HERE

## 📌 Resources

- [Security guidelines](https://github.com/nextauthjs/.github/blob/main/SECURITY.md)
- [Contributing guidelines](https://github.com/nextauthjs/.github/blob/main/CONTRIBUTING.md)
- [Code of conduct](https://github.com/nextauthjs/.github/blob/main/CODE_OF_CONDUCT.md)
- [Contributing to Open Source](https://kcd.im/pull-request)

### CODE REVIEW

**Comment by adamelmore** [comment]
gimme 😍

let me know if i can be helpful, very excited to add authjs to my astro project

**Comment by balazsorban44** [comment]
This is awesome! we have a few items on the backlog, but once we get around it, I'll be super happy to check this, out! Thanks!

File: packages/frameworks-astro/index.ts [hunk 1]
```diff
@@ -1,3 +1,33 @@
...
+ *   output: "server",
+ *   adapter: node({
+ *     mode: 'standalone'
+ *   })
+ * });
+ * ```
+ * :::
+ *
+ * ```bash npm2yarn2pnpm
+ * npm install @auth/core @auth/sveltekit
```
Comment (tajkirkpatrick) [inline packages/frameworks-astro/index.ts:25]: You didn't mean to put SvelteAuth

Comment (TheOtterlord) [reply packages/frameworks-astro/index.ts]: AHH, thanks for catching this :sweat_smile:

**Comment by kb1995** [comment]
This looks amazing - I was just building a new Astro project so will definitely be an early user to this

**Comment by TheOtterlord** [comment]
@kb1995 Thanks! If you want to try it now, we have an unofficial package [`auth-astro`](https://www.npmjs.com/package/auth-astro) to help us test and iterate on ideas before this is merged. We're working on a few upgrades at the moment which should make their way to that package very soon.

**Comment by kb1995** [comment]
@TheOtterlord Thanks a lot for this - the library is so easy to integrate. The github provider works flawlessly, but when I tried to set up Twitter, I got the following:

```
[auth][error][InvalidState]: Read more at https://errors.authjs.dev#invalidstate
[auth][cause]: InvalidState: State was missing from the cookies.
    at Object.use (file:///Users/kris/Documents/code/products/astro-personal-sidebar/node_modules/@auth/core/lib/oauth/checks.js:70:19)
    at handleOAuth (file:///Users/kris/Documents/code/products/astro-personal-sidebar/node_modules/@auth/core/lib/oauth/callback.js:44:38)
    at Module.callback (file:///Users/kris/Documents/code/products/astro-personal-sidebar/node_modules/@auth/core/lib/routes/callback.js:14:47)
    at AuthInternal (file:///Users/kris/Documents/code/products/astro-personal-sidebar/node_modules/@auth/core/lib/index.js:64:51)
    at async Proxy.Auth (file:///Users/kris/Documents/code/products/astro-personal-sidebar/node_modules/@auth/core/index.js:100:30)
    at async eval (/node_modules/auth-astro/index.ts:50:17)
    at async Module.get (/node_modules/auth-astro/index.ts:73:14)
    at async call (file:///Users/kris/Documents/code/products/astro-personal-sidebar/node_modules/astro/dist/core/endpoint/index.js:70:20)
    at async call (file:///Users/kris/Documents/code/products/astro-personal-sidebar/node_modules/astro/dist/core/endpoint/dev/index.js:15:10)
    at async handleRoute (file:///Users/kris/Documents/code/products/astro-personal-sidebar/node_modules/astro/dist/vite-plugin-astro-server/route.js:117:20)
[auth][details]: {
  "provider": "twitter"
}
```

I also noticed that when I have two sign in buttons

```
<SignIn provider="github">Sign in with Github</SignIn>
<SignIn provider="twitter">Sign in with Twitter</SignIn>
```

Only one will work. In this case, `Sign in with Github` attempted to sign in with Twitter, but the Twitter button didn't do anything.

My `[...astroauth].ts` looks like this:

```
import { AstroAuth, type AstroAuthConfig } from "auth-astro";
import GitHub from "@auth/core/providers/github";
import Twitter from "@auth/core/providers/twitter";

export const authOpts: AstroAuthConfig = {
  providers: [
    //@ts-expect-error issue https://github.com/nextauthjs/next-auth/issues/6174
    GitHub({
      clientId: import.meta.env.GITHUB_ID,
      clientSecret: import.meta.env.GITHUB_SECRET,
    }),
    //@ts-expect-error issue https://github.com/nextauthjs/next-auth/issues/6174
    Twitter({
      clientId: import.meta.env.TWITTER_ID,
      clientSecret: import.meta.env.TWITTER_SECRET,
    }),
  ],
};

export const { get, post } = AstroAuth(authOpts);
```

**Comment by TheOtterlord** [comment]
yeah, I'm working on a fix for the button issue, but I've not seen an issue with twitter yet. I'll try to replicate this next time I'm free. thanks for reporting these bugs :heart:

**Comment by kb1995** [comment]
@TheOtterlord This might be unrelated, but do you know where we can put the type extension within this article https://authjs.dev/reference/adapters/supabase#extend-the-session-type-with-the-supabaseaccesstoken within Astro in order to make it work? Putting it in the recommended folder doesn't work

**Comment by TheOtterlord** [comment]
Only a day late :sweat_smile: 

This latest change turns the package into an Astro integration and removes a large portion of the installation process, as well as remove some annoying bugs & DX decisions. I've also updated the documentation to rely on a `components.d.ts` file removing any definitions outside the package.

File: packages/frameworks-astro/src/components/SignIn.astro [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  /** The authentication provider to sign in with. */
+  provider?: LiteralUnion<BuiltInProviderType, string>
+  options?: SignInOptions
+  authParams?: SignInAuthorizationParams
+}
+
+const key = Math.random().toString(36).slice(2, 11)
+
+const { provider, options, authParams, ...attrs } = Astro.props
+attrs.class = `signin-${key} ${attrs.class ?? ''}`
```
Comment (tony-sull) [inline packages/frameworks-astro/src/components/SignIn.astro:16]: tiny nit: I usually recommend leaning on `class:list` when passing `class` props to Astro components

```html
---
const { class: className, ...attrs } = Astro.props
---

<button class:list={[`signin-${key}`, className]} {...attrs}>
    <slot />
</button> 
```

**Comment by nowaythatworked** [comment]
@tony-sull We are actually developing this further on this repo https://github.com/nowaythatworked/auth-astro and merge the results in here periodically. Feel free to PR or join the Discussion on the Astro Discord

**Comment by lgs** [comment]
this is bad !

**Comment by ChrisSki** [comment]
Any update on this? Waiting for this bad boy so I can get Auth working with Astro 🙏

**Comment by Dezaku** [comment]
> Any update on this? Waiting for this bad boy so I can get Auth working with Astro 🙏 

Same!

**Comment by TheOtterlord** [comment]
The current version of this PR is behind `auth-astro`, but if there's still interest, I'd be happy to update it when I get the time!

**Comment by AsyncBanana** [comment]
I also have fixed some things (monorepo functionality, weird edge bugs, consistent formatting) and added some stuff (defineConfig helper) in [my fork](https://github.com/AsyncBanana/auth-astro), if you want to take some or all of it. I have submitted some changes as PRs already, and would be happy to submit other changes or the whole thing as a PR.

**Comment by martincolladodev** [comment]
> Any update on this? Waiting for this bad boy so I can get Auth working with Astro 🙏 

Same!

**Comment by Dezaku** [comment]
We need an update asap!!

**Comment by Ayfri** [comment]
Any update ?

**Comment by brunocascio** [comment]
Really looking for this to be merged <3

**Comment by nathandaly** [comment]
I'm using this for commercial purposes but let's get this officially supported. Astro is huge.

**Comment by Gradlon** [comment]
Would be really could to get this merged!

**Comment by podbasico** [comment]
Hello, any update on this PR?

**Comment by blainehill2001** [comment]
+1 for Astro with Auth!!

**Comment by ndom91** [comment]
@TheOtterlord can you take a look at the merge conflicts? Thanks!

**Comment by TheOtterlord** [comment]
@ndom91 this PR fell behind both auth-astro and this repo, so I've re-implemented using the latest version of @auth/core and the docs in a new PR: https://github.com/nextauthjs/next-auth/pull/9856

Hope to continue the conversation over there! Excited to pick this up again!

**Pull request closed without merging**